modules/sharedIdSystem.js

```diff
@@ -0,0 +1,210 @@
+      }
+      callback(value);
+    },
+    error: function (statusText, responseBody) {
+      const value = constructCookieValue(sharedIdGenerator.id(), true);
+      utils.logInfo('SharedId: Ulid Generated SharedId: ' + value.id);
+      callback(value);
+    }
+  }
+}
+
+/**
+ * existing id generation call back
+ * @param result
+ * @param callback
+ * @returns {{success: success, error: error}}
+ */
+function existingIdCallback(storedId, callback) {
+  return {
+    success: function (responseBody) {
+      utils.logInfo('SharedId: id to be synced: ' + storedId.id);
+      if (responseBody) {
+        try {
+          let responseObj = JSON.parse(responseBody);
+          storedId = constructCookieValue(responseObj.sharedId, false);
+          utils.logInfo('SharedId: Older SharedId: ' + storedId.id);
+        } catch (error) {
+          utils.logError(error);
+        }
+      }
+      callback(storedId);
+    },
+    error: function () {
+      utils.logInfo('SharedId: Sync error for id : ' + storedId.id);
+      callback(storedId);
+    }
+  }
+}
+
+/**
+ * Encode the id
+ * @param value
+ * @returns {string|*}
+ */
+function encodeId(value) {
+  const result = {};
+  const sharedId = (value && typeof value['id'] === 'string') ? value['id'] : undefined;
+  if (sharedId == OPT_OUT_VALUE) {
+    return undefined;
+  }
+  if (sharedId) {
+    const bidIds = {
+      id: sharedId,
+    }
+    const ns = (value && typeof value['ns'] === 'boolean') ? value['ns'] : undefined;
+    if (ns == undefined) {
+      bidIds.third = sharedId;
+    }
+    result.sharedid = bidIds;
+    utils.logInfo('SharedId: Decoded value ' + JSON.stringify(result));
+    return result;
+  }
+  return sharedId;
+}
+
+/** @type {Submodule} */
+export const sharedIdSubmodule = {
+  /**
+   * used to link submodule with config
+   * @type {string}
+   */
+  name: MODULE_NAME,
+
+  /**
+   * decode the stored id value for passing to bid requests
+   * @function
+   * @param {string} value
+   * @returns {{sharedid:{ id: string, third:string}} or undefined if value doesn't exists
+   */
+  decode(value) {
+    return (value) ? encodeId(value) : undefined;
+  },
+
+  /**
+   * performs action to obtain id and return a value.
+   * @function
+   * @param {SubmoduleParams} [configParams]
+   * @returns {sharedId}
+   */
+  getId(configParams) {
+    const resp = function (callback) {
+      utils.logInfo('SharedId: Sharedid doesnt exists, new cookie creation');
+      ajax(ID_SVC, idGenerationCallback(callback), undefined, {method: 'GET', withCredentials: true});
+    };
+    return {callback: resp};
+  },
+
+  /**
+   * performs actions even if the id exists and returns a value
+   * @param configParams
+   * @param storedId
+   * @returns {{callback: *}}
+   */
+  extendId(configParams, storedId) {
+    utils.logInfo('SharedId: Existing shared id ' + storedId.id);
+    const resp = function (callback) {
+      const needSync = isIdSynced(configParams, storedId);
```

Comment:
couldn't you rely on the `refreshInSeconds` feature instead of reproducing it?

https://github.com/prebid/Prebid.js/blob/master/modules/userId/index.js#L64

Reply:
couple of reason for not using "refreshInSeconds"

1. The default of refreshInSeconds is the expiry of the cookies but we are looking at refreshing/syncing atleast once in 24 hours
2. The refresh based on refreshInSeconds relies on getId function to reset where as we want to use extendId function

Reply:
@SKOCHERI ok. i'll note that I still think there are better ways to accomplish this (some of which may require userId module changes), but I think it's fine to leave as is and perhaps gets reviewed at another time down the road.